### DISCUSSION TITLE
Using the json grammar with ./server and Python

### DISCUSSION THREAD

**Discussion by IridiumMaster** [ggml-org/llama.cpp#3268] [Q&A] [answered]
Hello,
Sorry to trouble you, but I have been a little confounded by how to get ./server to parse any of the grammars that are provided as examples with llama.cpp. 
I started by passing the json.gbnf file from grammars in as a string. 
`def run_prompt(self, prompt, grammar, my_preset_rotation=0, max_tokens=3000, max_retries=1, timeout=240):

        headers = {
            'Content-Type': 'application/json'
        }

        payload = {
            'prompt': prompt,
            'n_predict': max_tokens,
            'temperature': 0.7,
            'top_p': 0.9,
            'repeat_penalty': 1.15,
            'grammar': grammar  # include grammar here
        }

        with httpx.Client() as client:
            response = client.post(self.URI, headers=headers, json=payload, timeout=timeout)

        if response.status_code == 200:
            logString = response.json()
            # logging.info(logString)
            logging.info("***")
            result = response.json()['content']
            logging.info(result)
            return result
        else:
            logging.info("Error")
            logString = str(response.status_code)
            logging.info(logString)
            logging.info(response.text)
            return "#Error#, " + str(response.status_code) + ", " + response.text`

This caused all sorts of parsing errors on the server side and failed with a 400 status. I then switched to sending the grammar file as a payload, like:
` payload = {
            'prompt': prompt,
            'n_predict': max_tokens,
            'temperature': 0.7,
            'top_p': 0.9,
            'repeat_penalty': 1.15,
        }

        files = {'grammar': open(grammar_file_path, 'rb')}

        with httpx.Client() as client:
            response = client.post(self.URI, headers=headers, data=payload, files=files, timeout=timeout)`

This also failed with various server side errors. Could any of you help me figure out what is going wrong? I really appreciate your assistance.

Update, so it appears that I can get this to work by pasting in the encoded string from the frontend sample:
`'grammar': 'root   ::= object\nvalue  ::= object | array | string | number | (\"true\" | \"false\" | \"null\") ws\n\nobject ::=\n  \"{\" ws (\n            string \":\" ws value\n    (\",\" ws string \":\" ws value)*\n  )? \"}\" ws\n\narray  ::=\n  \"[\" ws (\n            value\n    (\",\" ws value)*\n  )? \"]\" ws\n\nstring ::=\n  \"\\\"\" (\n    [^\"\\\\] |\n    \"\\\\\" ([\"\\\\/bfnrt] | \"u\" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F]) # escapes\n  )* \"\\\"\" ws\n\nnumber ::= (\"-\"? ([0-9] | [1-9] [0-9]*)) (\".\" [0-9]+)? ([eE] [-+]? [0-9]+)? ws\n\n# Optional space: by convention, applied in this grammar after literal chars when allowed\nws ::= ([ \\t\\n] ws)?\n'
` It looks like I need to do some work to figure out why httpx is not auto encoding this properly

**Comment by ianscrivener** [comment, marked as answer]
There is some good information on grammer in these discussions: [2494](https://github.com/ggerganov/llama.cpp/discussions/2494), [2560](https://github.com/ggerganov/llama.cpp/discussions/2560), [1773](https://github.com/ggerganov/llama.cpp/pull/1773) and [2637](https://github.com/ggerganov/llama.cpp/discussions/2637) 
 
 Have you seen the [Grammer Builder](https://grammar.intrinsiclabs.ai/)?
